packages/vats/src/lib-board.js

```diff
@@ -60,81 +172,118 @@ function makeBoard(
+     *
+     * Each board ID includes a CRC so that the last part is well-distributed,
+     * which mitigates typo risks.
+     *
+     * Scaling Consideration: since we cannot know when consumers
+     * are no longer interested in an ID, a board holds a strong reference
+     * to `value` for its entire lifetime. For a well-known board, this
+     * is essentially forever.
+     *
+     * @param {Key} value
+     * @throws if `value` is not a Key in the @agoric/store sense
+     */
+    getId(value) {
+      return getId(value, this.state);
     },
-    getValue: id => {
-      assert.typeof(id, 'string', X`id must be string: ${id}`);
-      id.startsWith(prefix) || Fail`id must start with ${prefix}: ${id}`;
-      const digits = id.slice(prefix.length);
-      digits.match(DIGITS_REGEXP) ||
-        Fail`id must end in at least ${q(crcDigits + 1)} digits: ${id}`;
-      const seq = digits.slice(crcDigits);
-      const allegedCrc = digits.slice(0, crcDigits);
-      const crcInput = `${prefix}${seq}`;
-      const crc = calcCrc(crcInput, crcDigits);
-      allegedCrc === crc ||
-        Fail`id is probably a typo, cannot verify CRC: ${id}`;
-      idToVal.has(id) || Fail`board does not have id: ${id}`;
-      return idToVal.get(id);
+    /**
+     * @param {string} id
+     * @throws if id is not in the mapping
+     */
+    getValue(id) {
+      return getValue(id, this.state);
     },
-    // Adapter for lookup() protocol.
-    lookup: async (...path) => {
+    /**
+     * Adapter for NameHub lookup() protocol.
+     *
+     * @param  {...string} path
+     */
+    async lookup(...path) {
+      const { self } = this;
       if (path.length === 0) {
-        return board;
+        return self;
       }
       const [first, ...rest] = path;
-      const firstValue = board.getValue(first);
+      const firstValue = self.getValue(first);
       if (rest.length === 0) {
         return firstValue;
       }
       return E(firstValue).lookup(...rest);
     },
-    has: valToId.has,
-    ids: () => harden([...idToVal.keys()]),
+    /** @param {Key} val */
+    has(val) {
+      const { state } = this;
+      return state.valToId.has(val);
+    },
+    ids() {
+      const { state } = this;
+      return harden([...state.idToVal.keys()]);
+    },
+    /**
+     * Get a marshaller that implements the convertValToSlot hook using getId(),
+     * "publishing" previously un-mapped objects.
+     */
+    getPublishingMarshaller() {
+      return makeMarshallers(this.state).publishingMarshaller;
+    },
+    /**
+     * Get a marshaller that implements the convertValToSlot hook using getId(),
+     * only for objects that the board has().
+     *
+     * For other objects, we don't throw, but we emit a null slot.
+     * Un-serializing a null slot always produces a fresh object
+     * rather than preserving object identity.
+     */
+    getReadonlyMarshaller() {
+      return makeMarshallers(this.state).readonlyMarshaller;
```

Comment:
If each request to `getReadonlyMarshaller()`, what happens to all those objects? @dckc when can they be GCed?

trying to figure out whether we need a cache